src/huggingface_hub/hf_api.py

```diff
@@ -264,6 +266,122 @@ def delete_repo(
+
+            repo_id (:obj:`str`):
+                The repository to which the file will be uplaoded, for example: :obj:`"username/custom_transformers"`
+
+            repo_type (:obj:`str`, Optional):
+                Set to "dataset" if uploading to a dataset, default is model.
+
+            revision (:obj:`str`, Optional):
+                The git revision to commit from. Defaults to the :obj:`"main"` branch.
+
+            token (:obj:`str`, Optional):
+                Authentication token, obtained with :function:`HfApi.login` method
+
+        Returns:
+            :obj:`str` : The URL to visualize the uploaded file on the hub
+
+        Raises:
+            :class:`ValueError` : if some parameter value is invalid
+
+            :class:`requests.HTTPError` : if the HuggingFace API returned an error
+
+        Examples:
+            >>> with open("./local/filepath", "rb") as fobj:
+            ...     upload_file(
+            ...         path_or_fileobj=fileobj,
+            ...         path_in_repo="remote/file/path.h5",
+            ...         repo_id="username/private-dataset",
+            ...         repo_type="datasets",
+            ...         token="my_token",
+            ...    )
+            "https://huggingface.co/datasets/username/private-dataset/blob/main/remote/file/path.h5"
+
+            >>> upload_file(
+            ...     path_or_fileobj=".\\\\local\\\\file\\\\path",
+            ...     path_in_repo="remote/file/path.h5",
+            ...     repo_id="username/private-model",
+            ...     token="my_token",
+            ... )
+            "https://huggingface.co/username/private-model/blob/main/remote/file/path.h5"
+
+
+        """
+        if repo_type not in REPO_TYPES:
+            raise ValueError("Invalid repo type, must be one of {}".format(REPO_TYPES))
+
+        # Validate path_or_fileobj
+        if isinstance(path_or_fileobj, str):
+            path_or_fileobj = os.path.normpath(os.path.expanduser(path_or_fileobj))
+            if not os.path.isfile(path_or_fileobj):
+                raise ValueError(
+                    "Provided path: '{}' is not a file".format(path_or_fileobj)
+                )
+        elif not isinstance(path_or_fileobj, (RawIOBase, BufferedIOBase)):
+            # ^^ Test from: https://stackoverflow.com/questions/44584829/how-to-determine-if-file-is-opened-in-binary-or-text-mode
+            raise ValueError(
+                "path_or_fileobj must be either an instance of str or BinaryIO. "
+                "If you passed a fileobj, make sure you've opened the file in binary mode."
+            )
+
+        # Normalize path separators and strip leading slashes
+        path_in_repo = normalize(path_in_repo)
+        if os.path.isabs(path_in_repo):
+            raise ValueError(
+                "path_in_repo must be a relative path, but '{}' is an absolute path".format(
+                    path_in_repo
+                )
+            )
```

Comment:
maybe better to require a filename with the right format (with forward slashes, etc.) instead of trying to normalize it?

Reply:
Agreed!